src/main/BlockPool.cpp

```diff
@@ -0,0 +1,863 @@
+Superblock* small_superblocks[NUM_SMALL_POOLS];
+
+// Constant indicating a deleted bucket.
+uintptr_t deleted_bucket = UINTPTR_MAX;
+
+unsigned add_collision = 0;
+unsigned remove_collision = 0;
+unsigned lookup_collision = 0;
+
+// Find the 2-log of the next power of two greater than or equal to size.
+static unsigned next_log2_32(int size) {
+  if (size == 0) {
+    return 0;
+  }
+  int log2 = 0;
+  int temp = size;
+  if (temp & 0xFFFF0000) {
+    log2 += 16;
+    temp >>= 16;
+  }
+  if (temp & 0xFF00) {
+    log2 += 8;
+    temp >>= 8;
+  }
+  if (temp & 0xF0) {
+    log2 += 4;
+    temp >>= 4;
+  }
+  if (temp & 0xC) {
+    log2 += 2;
+    temp >>= 2;
+  }
+  if (temp > 0) {
+    log2 += temp - 1;
+  }
+  if ((size & (1 << log2)) && (size & ((1 << log2) - 1))) {
+    log2 += 1;
+  }
+  return log2;
+}
+
+// Arena is 1Gb = 30 bits.
+#define ARENASIZE (1 << 30)
+
+#define SUPERBLOCK_BITS (18)
+#define SUPERBLOCK_SIZE (1 << SUPERBLOCK_BITS)
+
+// NOTE: We use a fixed superblock header size, regardless of block size. This
+// leaves some unused bitset entries for larger block sizes.
+#define SUPERBLOCK_HEADER_SIZE (1040)
+
+// 1 Mbyte for medium block arenas.
+#define LARGE_ARENA_SIZE_LOG2 (19)
+#define LARGE_ARENA_SIZE (1 << LARGE_ARENA_SIZE_LOG2)
+#define LARGE_HEADER_SIZE (1040)
+
+void* superblock_arena = nullptr;
+uintptr_t arena_superblock_start;
+uintptr_t arena_superblock_end;
+uintptr_t arena_superblock_next;
+
+// Allocates a new superlbock from the small object arena.  If the arena is
+// full, we call abort with the message "out of superblock space".
+Superblock* new_superblock(int block_size) {
+  if (arena_superblock_next >= arena_superblock_end) {
+    allocerr("out of superblock space");
+  }
+  void* pointer = reinterpret_cast<void*>(arena_superblock_next);
+  unsigned superblock_size =
+      (SUPERBLOCK_SIZE - SUPERBLOCK_HEADER_SIZE) / block_size;
+  unsigned bitset_entries = (superblock_size + 63) / 64;
+  Superblock* superblock = new (pointer)Superblock(block_size, bitset_entries);
+  arena_superblock_next += SUPERBLOCK_SIZE;
+  return superblock;
+}
+
+// Compute the address of the superblock header from a pointer.
+// Returns nullptr if candidate pointer is not inside the small object arena
+// or if the pointer points inside the superblock header.
+Superblock* superblock_from_pointer(void* pointer) {
+  uintptr_t candidate = reinterpret_cast<uintptr_t>(pointer);
+  if (candidate >= arena_superblock_start && candidate < arena_superblock_next) {
+    if ((candidate & (SUPERBLOCK_SIZE - 1)) < SUPERBLOCK_HEADER_SIZE) {
+      // The pointer points inside the superblock header.
+      return nullptr;
+    }
+    return reinterpret_cast<Superblock*>(
+        candidate & ~uintptr_t{SUPERBLOCK_SIZE - 1});
+  } else {
+    return nullptr;
+  }
+}
+
+// Free a pointer inside a given superblock. The block MUST be in the given
+// superblock.
+void free_small_block(void* pointer, Superblock* superblock) {
```

Comment:
Is there any reason why this isn't a method in Superblock?  (and similarly for some of the other functions)